geoutils/georaster.py

```diff
@@ -822,10 +913,179 @@ def format_value(value):
+        :param j: column (j) index of pixel.
+        :type j: array-like
+        :param offset: return coordinates as "corner" or "center" of pixel
+        :param offset: str
+
+        :returns x, y: x,y coordinates of i,j in reference system.
+        """
+
+        x,y = self.ds.xy(i,j,offset=offset)
+
+        return x, y
+
+    def outside_image(self, xi,yj, index=True):
+        """
+        #TODO: calculate matricially for all points instead of doing for only one?
+        Check whether a given point falls outside of the raster.
+
+        :param xi: Indices (or coordinates) of x direction to check.
+        :param yj: Indices (or coordinates) of y direction to check.
+        :param index: Interpret ij as raster indices (default is True). If False, assumes ij is coordinates.
+        :type xi: array-like
+        :type yj: array-like
+        :type index: bool
+
+        :returns is_outside: True if ij is outside of the image.
+        """
+        if not index:
+            xi,xj = self.xy2ij(xi,yj)
+
+        if np.any(np.array((xi,yj)) < 0):
+            return True
+        elif xi > self.ds.width or yj > self.ds.height:
+            return True
+        else:
+            return False
+
+    def interp_points(self,pts,nsize=1,mode='linear',band=1):
+
+        """
+        Interpolate raster values at a given point, or sets of points.
+
+       :param pts: Point(s) at which to interpolate raster value. If points fall outside of image,
+       value returned is nan.'
+       :param nsize: Number of neighboring points to include in the interpolation. Default is 1.
+       :param mode: One of 'linear', 'cubic', or 'quintic'. Determines what type of spline is
+           used to interpolate the raster value at each point. For more information, see
+           scipy.interpolate.interp2d. Default is linear.
+       :param band: Raster band to use
+       :type pts: array-like
+       :type nsize: int
+       :type mode: str
+       :type band: int
+
+       :returns rpts: Array of raster value(s) for the given points.
+       :rtype rpts: array-like
+       """
+        assert mode in ['mean', 'linear', 'cubic', 'quintic',
+                        'nearest'], "mode must be mean, linear, cubic, quintic or nearest."
+
+        rpts = []
+
+        #TODO: might need to check if coordinates are center or point in the metadata here...
+
+
+        xx, yy = self.xxyy(offset='center', grid=False)
+        #TODO: right now it's a loop... could add multiprocessing parallel loop outside,
+        # but such a method probably exists already within scipy/other interpolation packages?
+        for pt in pts:
+            i,j = self.xy2ij(pt[0],pt[1])
+            if self.outside_image(i,j, index=True):
+                rpts.append(np.nan)
+                continue
+            else:
+                x = xx[j - nsize:j + nsize + 1]
+                y = yy[i - nsize:i + nsize + 1]
+
+                #TODO: read only that window?
```

Comment:
Here, we could provide the entire list to griddata to get all points derived simultaneously. Or we could write the for loop the same way Amaury did and interpolate only in a window read by rio to optimize RAM (and add a multiprocessing option to the for loop?)

Reply:
Maybe we provide both options? And should we merge value_at_coords and interp_points then?

Reply:
In georaster, value_at_coords and interp have different use. value_at_coords will only extract the raster at certain pixel locations, for example to analyze the value at a pixel or a window around that pixel (I don't think I've ever used, that was more for Andrew I believe). On the other hand, interp() is meant to interpolate the value at any (including non integer) pixel locations. 
We could potentially merge both, but I see some differences that also support keeping them separate:
- value_at_coords read the data directly at requested location. For large arrays that's a useful feature
- on the other hand, interp loads the entire data set for the interpolation (that answers your first question).
So for now, I would keep them separate as it is, but you need to fix the issue that for now only integer pixels are read. For the interpolation, you could use ndimage.map_coordinates like we did in georaster's interp: https://github.com/GeoUtils/georaster/blob/0df19bfd91fd10e92beb54308428e212e1d19c1a/georaster/georaster.py#L936